system/lib/libc/dynlink.c

```diff
@@ -148,6 +152,178 @@ static void ensure_init() {
+void emscripten_promise_reject(int promise_id);
+
+void _emscripten_thread_sync_code_async(struct dso* dso, int promise_id);
+
+static void sync_one_onsuccess(struct dso* dso, void* user_data) {
+  int promise_id = (intptr_t)user_data;
+  // Load the next dso in the list
+  thread_local_tail = dso;
+  _emscripten_thread_sync_code_async(dso->next, promise_id);
+}
+
+static void sync_one_onerror(struct dso* dso, void* user_data) {
+  int promise_id = (intptr_t)user_data;
+  pthread_rwlock_unlock(&lock);
+  emscripten_promise_reject(promise_id);
+}
+
+void _emscripten_thread_sync_code_async(struct dso* dso, int promise_id) {
+  _emscripten_errf("_emscripten_thread_sync_code_async %p promise_id=%d", dso, promise_id);
+  ensure_init();
+  pthread_rwlock_rdlock(&lock);
+  if (!thread_local_tail) {
+    thread_local_tail = head;
+  }
+  if (!dso) {
+    dso = thread_local_tail->next;
+  }
+  if (!dso) {
+    pthread_rwlock_unlock(&lock);
+    emscripten_promise_resolve(promise_id, NULL);
+    return;
+  }
+
+  // Unlock happens once all DSO have been loaded, or one of them fails
+  // with sync_one_onerror.
+  _emscripten_dlopen_js(
+    dso, sync_one_onsuccess, sync_one_onerror, (void*)promise_id);
+}
+
+// Called on background threads to synchronously "catch up" with all the DSOs
+// that are currently loaded.
+bool _emscripten_thread_sync_code() {
+  // Should only ever be called from a background thread.
+  assert(!emscripten_is_main_runtime_thread());
+  ensure_init();
+  if (thread_local_tail == tail) {
+#ifdef DYLINK_DEBUG
+    _emscripten_errf("%p: emscripten_thread_sync_code: already in sync", pthread_self());
+#endif
+    return true;
+  }
+  pthread_rwlock_rdlock(&lock);
+  if (!thread_local_tail) {
+    thread_local_tail = head;
+  }
+  while (thread_local_tail->next) {
+    struct dso* p = thread_local_tail->next;
+#ifdef DYLINK_DEBUG
+    fprintf(stderr,
+            "%p: emscripten_thread_sync_code: %s mem_addr=%p mem_size=%zu "
+            "table_addr=%p table_size=%d",
+            pthread_self(),
+            p->name,
+            p->mem_addr,
+            p->mem_size,
+            p->table_addr,
+            p->table_size);
+#endif
+    void* success = _dlopen_js(p);
+    if (!success) {
+      _emscripten_errf("emscripten_thread_sync_code failed: %s", dlerror());
+      return false;
+    }
+    thread_local_tail = p;
+  }
+  pthread_rwlock_unlock(&lock);
+#ifdef DYLINK_DEBUG
+  _emscripten_errf("%p: emscripten_thread_sync_code done", pthread_self());
+#endif
+  return true;
+}
+
+// This structure exists so that we can stash the status of work we do
+// during emscripten_proxy_async_with_callback.
+// If the function that emscripten_proxy_async_with_callback ran included
+// as return value we coudl avoid this struct and the malloc/free of it.
+// See https://github.com/emscripten-core/emscripten/issues/18378
+struct thread_sync_data {
+  int promise_id;
+  bool result;
+};
```

Comment:
@tlively WDYT?

Reply:
If we supported a void* parameter to the callback, you would have to pack both `promise_id` and `result` into that pointer to avoid an allocation, right?